You can smuggle smuggler's satchels to a future shift

Incoming5643 opened the pull request:
:exclamation: YOUR FIRST INSTINCT MAY BE TO THINK THIS IS OP TRASH, BUT READ MY COMMENTS, IT'S NOT :exclamation: 

Hidden Smuggler Satchels hidden beneath the station now persist between rounds, though there's no assurances of WHEN they'll appear next and only one will ever appear in any given round.

Cavets: 
*Only one random item in a satchel is saved.
*The more satchels are hidden, the less likely any one satchel is to be in a round. At the very least, there has to be 20 satchels in rotation to have one appear in the next round, so at the very best, your odds of "reclaiming" a satchel are 5%
*Satchels are bound to the maps in which they are hidden, and only on the station z level
*It's a path save, so things you put together yourself like TTV or things with reagent components will only spawn in as their vanilla bits.
*Admin shenanigans can't be saved.
*Certain very high risk items can be replaced with less dangerous look alikes on a case by case basis

:cl:
add: Smuggler satchels left behind hidden on the station can now reappear in a later round.
experiment: There's absolutely no way of knowing when exactly it will reappear however...
add: Only one item in the satchel will remain when the satchel reappears.
add: Items are saved in their <b>initial</b> forms, so saving things like chemical grenades will only disappoint you.
add: The item pool is map specific and needs to be a certain size before satchels can start reappearing, so get to hiding that loot!
/:cl:

Incoming5643 committed:
Hidden Smuggler Satchels hidden beneath the station now persist between rounds, though there's no assurances of WHEN they'll appear next and only one will ever appear in the round.

NOTE THAT THIS PULL INTRODUCES A CRITICAL BUG, DO NOT MERGE *JUST* THIS PR, AWAIT THE FOLLOWUP

francinum commented:
OH GOD NO

Incoming5643 commented:
So essentially when CollectSecretSatchels() is called (either normally or manually) if there are any actual satchels to document, the game locks up completely and the save file grows to infinity.

This happens before any of the debug world calls trigger. I am stumped.

GunHog commented:
I am so going to put bombs into them.

Incoming5643 commented:
It's only a path save, the best you could do is a minibomb or a disembodied bomb core. A TTV would only give you the valve, a chem grenade wouldn't contain anything.

GunHog commented:
Okay, maybe I can drop a facehugger in there, then. lol

Shadowlight213 commented:
>even saves locations

This is such a horrible idea. seriously. Persistence memes have gotten out of hand.
Fuckit. Why not do persistent inventory of everyone who escapes next?

Incoming5643 commented:
You never really know until you try. The built in heavy variance in when a satchel might reappear means that people can't just bank on going back for their stuff when they get a traitor roll. Even if you do find something great, you might not actually be in a position to make much use of it.

Final pull will have a line the prevents any satchel from spawning until there's a healthy number of hidden satchels to choose from, it's just missing from this pull because I took it out for testing.

Jordie0608 commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -0,0 +1,60 @@
+	secret_satchels << old_secret_satchels
+	var/path = text2path(chosen_satchel["object"])
+	var/obj/item/weapon/storage/backpack/satchel/flat/secret/F = new(chosen_satchel["loc"])
+	new path(F) //runtimes intentionally if satchel item has had its path changed/removed
```
I don't really like the idea of "runtimes intentionally"

Incoming5643 replied:
Is there a more elegant way to check to see if a path is valid that doesn't provoke a runtime? I know about `try` but the way it works doesn't really seem like it'd prevent the runtime.

Comment is mostly a meme placeholder while I look for a better solution

optimumtact replied:
there is an ispath check is there not?

if(ispath(path))

Incoming5643 replied:
ispath actually checks for validity? I thought it would accept anything that looked properly formated.

optimumtact replied:
I dunno, I hoped it would.

MrStonedOne replied:
Is path don't care if the type exists. Text2type might

Incoming5643 replied:
text2path will actually return null if the type doesn't exist, so ironically I already solved this problem.

Cheridan commented:
What happens if I place a new tile in space and then hide the satchel under it

Incoming5643 commented:
If it's off the z level it's never saved.

If it's on the z level during some future round the satchel will spawn in space where the tile had been, and will not be saved again if it's not recovered during that round since it won't be hidden anymore.

TechnoAlchemisto commented:
So given that very few people use smugglers satchels, If I am the only traitor to hide one, and nobody has before me, is it guaranteed it will be there the next round?

PKPenguin321 commented:
yeah honestly this may actually be a bit far for persistence features
i guess time will tell but now that i think about it i cant really think of too many positives that will come from this. a traitor stashes his ebow for next round and goes out and validhunts? a facehugger by some miracle appears in the same round as a smuggler satchel and gets reused the next round? a sneaky traitor steals a pusle rifle from a meme round death squad and holds onto it for later? it's pretty sketchy

PKPenguin321 commented:
also it completely violates my headcanon that the station is a different replica each time

AnturK commented:
Not a fan of adding even more persistence features. New round should be a clean slate in my opinion. And with the restrictions on this you might as well just generate few random satchels instead.

GunHog commented:
Just a note that this somewhat overlaps with the Black Box reward.

Bawhoppen commented:
This is dumb for alot of reasons (ignoring balance) Incoming how could you

Incoming5643 commented:
@TechnoAlchemisto With current testing code, yes. Final code would keep any satchels from spawning until a good number of them were in the pool, so you couldn't ever reliably regain any specific item, just a more or less random one that someone in a previous round deemed valuable/funny to save. 

---

Don't know how many times I have to stress the "This is a horribly inconsistent way to regain your own traitor gear unless you're literally the only player on the server" angle.  Consider:

1. There's only ever one satchel
2. That one satchel only has a 1 in [number of hidden satchels] chance of appearing (where [number of hidden satchels] can be arbitrarily large). There is no assured timescale for when your satchel will roll around again, theoretically it could take months.
3. When it does decide to spawn you have to actually be playing the game
4. Even if it's there you have to be the first one to get back there with a crowbar (remember, anyone with a T-ray can uncover hidden satchels)
5. Even then you would have needed to succeed your antag roll to actually make use of the thing you stowed

This is far less overpowered than the metastation throwing stars box

MMMiracles commented:
Never forget pun-pun bursting near round-start because someone put lamarr on him the round before.

ExcessiveUseOfCobblestone commented:
I wouldn't mind A Randomly Generated Smuggler Satchel on the station with neat items [hahaha people crowbarring the entire station for the possibility of a smuggler satchel sounds hilarious], but knowing I can stash say an Emag in there then check every round after in the same S E C R I T location until I get that emag [bonus if I'm antag] just seems kinda BS

Incoming5643 commented:
If you just make it a pool of items it'll get old the same way the maint loot table did. Leaving what is saved to players is pretty much the only way I can think of to keep things fresh.

callanrocks commented:
> put an uplink implant in one
> wait

Incoming5643 commented:
Uplinks won't preserve between rounds.

Don't know how many times I gotta say "this isn't a consistent means of item retrieval".

MMMiracles commented:
Pretty positive he means the actual uplink implant, which is an implant that gives you access to a 10TC uplink.

Incoming5643 commented:
Oh well yes that's pretty creative.

MMMiracles commented:
I would just have implanters put inside a satchel turned into generic empty ones as they'd be the worst offenders. Persistence stuff is kinda cool and all but the thought of shit like cargo stashing away bags of 20TC uplink implants for some future crew to find is actually terrifying.

Incoming5643 commented:
It only saves one item, if you put a container in the bag, it saves the container itself, not what's in it.

Jalleo commented:
Uh then I have noticed a issue do DS/ERT backpacks have hardcoded stuff to spawn in them? Or is that defined outside of their backpack?

Incoming5643 commented:
moot point since all backpacks are too big to fit in a smuggler's satchel

bgobandit commented:
this is going to be a fucking bitch to administrate given the high possibility but low provability of metagaming

lzimann commented:
let's stop with persistance memes

MrStonedOne commented:
I like it. I love the idea of somebody crowbarring up a tile and finding some random satchel some tator left behind.

Stop seeing it as presistance memes, and see it as crowd funded random station content.

Incoming5643 commented:
The sad thing is I only opened this pull request to get some help with a bug I was having, and literally 100% of the comments from others has been about balancing.

Anyway yeah work to be done. Also considering offering a seeded save file with like 50 or so satchels possible to start with so we can skip that creeping "start" phase.

MrStonedOne commented:
did you ever figure out that bug?

MrStonedOne commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -0,0 +1,60 @@
+		var/list/savable_obj = list()
+		for(var/obj/O in F)
+			if(ispath(O,/obj/item/stack/tile/plasteel) || ispath(O,/obj/item/weapon/crowbar) || O.admin_spawned)
+				continue
```
change to a typecache (see the black box for more info) much much faster.

MrStonedOne commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -0,0 +1,60 @@
+		if(savable_obj.len < 1)
+			continue
+		world << "<span class='boldannounce'>and it's valid!</span>"
+		saved_satchels += list("map" = MAP_NAME, "loc" = F.loc, "object" = pick(savable_obj))
```
the loc bit is what is causing your bug, save the xyz

Incoming5643 replied:
Rogered

Incoming5643 commented:
I haven't actually touched the branch since I opened this ¯\_(ツ)_/¯

It's the strangest thing, the save locks up before it even hits CollectSecretSatchels but it locks up because the save file grows to infinity (which only happens when CollectSecretSatchels is called). Some weird shenanigan with how saves work and/or I'm just missing something stupid obvious

MrStonedOne commented on code/controllers/subsystem/ticker.dm:
```diff
@@ -451,6 +451,8 @@ var/datum/subsystem/ticker/ticker
 		world.log << dellog
 
+	//Collects persistence features
+	SSpersistence.CollectData()
```
We need a modular way to do this.

Incoming5643 replied:
The end of round isn't very clean in terms of calling things, right now you either add something in round end code or you set your atom to processing and check for ticker.current_state == GAME_STATE_FINISHED.

The idea of making this a subsystem is to eventually unite all the persistence <s>memes</s> features under it.

bgobandit commented:
ideally it is crowd funded station content but in practice it is going to turn into people putting together stashes for themselves and then '''''just happening''''' to stumble across them in later rounds

like nethack and bones files 

the only way I can think of it working is if the location is randomized

Incoming5643 committed:
Finishes up satchel persistence.

Satchels require at least 20 saved bags before any will start spawning.

the var persistence_replacement can be set to a path to replace a very high risk item with a lower risk look alike. This is currently done with the nuke disk, the wizards spellbook, and the uplink implant

Incoming5643 commented:
Feature complete :heart: 

Reread the description, I've scaled things further back a bit.

RemieRichards commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -20,41 +21,46 @@ var/datum/subsystem/persistence/SSpersistence
-		satchels += potential_satchel
-	if(satchels.len < 1)
+	secret_satchels = new /savefile("data/npc_saves/SecretSatchels_[MAP_NAME].sav")
+	satchel_blacklist = typecacheof(/obj/item/stack/tile/plasteel) + typecacheof(/obj/item/weapon/crowbar)
```
```typecacheof(list(/obj/item/stack/tile/plasteel, /obj/item/weapon/crowbar))```?

RemieRichards commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -5,8 +5,9 @@ var/datum/subsystem/persistence/SSpersistence
 	var/savefile/secret_satchels
-	var/list/new_secret_satchels = list() //these are objects
-	var/list/old_secret_satchels = list() //these are associate
+	var/list/satchel_blacklist 		= list() //these are typecaches
```
technically the list is one ENTIRE typecache (singular, like how it's a list not a lists) but I'm not anal enough to ask you to change this (but you should change the other comment so maybe bundle it as a 2for1 :P )

Incoming5643 committed:
DESIGNATED.
REMIE.
PULL.

optimumtact commented:
Okay maybe it's just me, but I don't get it - in it's current watered down state I see no functional difference between what is currently here and a randomly spawning smugglers satchel with a list of random loot drops that be in it.

The problem is that the functionality you want (leaving items between rounds) harms the balance of our game (everyone starts on certain terms and can level up over the round so it has to be heavily nerfed to the point where it no longer makes any noticeable difference between it and a random maint lootdrop before players and contributors are comfortable with it.

If you ask me that's a pretty good sign that the feature as it currently is simply doesn't fit and this is the same friction that I assume the blackbox suffers from.

I don't know if there would ever be a case where a feature like this would ever work in a round based game as too many balance decisions are made with the assumption that everyone starts with a known quantity of items based on their role in the round. Any disruption to that causes everyone to complain about unfairness or balance.

Incoming5643 commented:
A drop list, ANY drop list, suffers from diminishing returns over time. Eventually even if you don't look up the code you will know exactly what is possible from said pool and subsequently either ignore the items or specifically rush to find "the best" item in the pool.

The black box was dangerous because on basil only a small handful of people fought the colossus, meaning that they could build up a huge array of rare and powerful gear, and there was no appreciable limit to how much you could store like this. This doesn't have that problem.

At the end of the day one item is just that, one item. Even if its an exceptionally powerful item (and it won't always be, I am fully expecting to see a lot of joke satchels get put down from surplus crates) it is just one thing. I've introduced a mechanic to allow the edge cases where "it's just one item" isn't a defensible point, but even crazy wizard shit is generally self limiting. Nothing you can fit in the satchel is a game winner by itself and there's absolutely no way to assure that you will get it anyway.

optimumtact commented:
But unless you rip up every single tile on the station you're unlikely to find any satchel anyway, so you really only have a random chance to find your own satchel?

The problem is that any nerf to make this palatable and not game breaking is going to be a nerf that simply renders it almost pointless, i.e no guarantee you will actually ever see your own smugglers satchel again, or find anyone elses.

You can say you fully expect to see lots of joke satchels put down, but again, no guarantee and with our playerbase I find it more likely they're only going to be interested in saving a single powerful item through each round.

I just don't understand how this is any different from a random lootspawn other than the items that go in the list are player defined and I don't see that being such a big selling point other than making sure you only put the one item you want to respawn in it, which also raises the question of why even bothering to have multiple slots?

Incoming5643 commented:
T-Ray scanners can reveal hidden satchels. Also yes, I assure you bored greytiders will start ripping up the tiling and it will be wonderful.

I believe in the power of pranks.

Satchels still have in round use for them so just having the one slot would be a big nerf. And yes a loot pool that can't go stale is a big selling point.

optimumtact commented:
I'm not saying you need to nerf them more. I'm saying with their current set of nerfs they're so weak as to basically match a random lootspawn in which case I don't see the point.

optimumtact commented:
The feedback on this is mixed and there are people I respect on both sides of the argument, I'm leaning towards a merge and see how we go

@Cheridan @KorPhaeron final signoff?

Incoming5643 commented:
If it turns out to consistently be overpowered trash I won't defend it.

Just keep in mind that since that the slow progression of how this pool builds itself means that we probably won't really know if that's true for months. You can't just open the revert the second someone manages to hide a pulse rifle under ian's bed.

TechnoAlchemisto commented:
Worth a try seems like it could be fun

KorPhaeron commented:
May as well try it, worst that happens is we revert

AnonymousNow commented:
Sounds rad.

Ninja137 commented:
So, this is a pretty decently far corner case, but I figure it's worth mentioning. Assuming that someone decided to dedicatedly attempt this, you could nab either 20 satchels per round, hide stolen gear in them, and pray.

Or you could go full meme, get a Crypto, 14 briefcases of cash, use mining to get plasma and sell a shitload of it to CentComm, buy 20+ Null Crates, get satchels off of those, load them all up with the contents of the crates (Revolvers, pistols, minibombs, implants, cryptos, all sorts of shit.), and go wild with it.

Neither one is really all that great, and the entire thing is basically pointless because it's 100% random, and the entire thing is just a bad idea, but still.

If it wasn't limited to only Smuggler's Satchels, or you could make Smuggler's Satchels as a non-traitor (Like, say, through sheets, the biogen, or some other method), then it'd probably be decently interesting. That way you'd have more than just traitor shit. You'd get people fucking around with notes, hiding drinks, or other things at least. And if there were multiple possible hidden caches per round, and maybe a number of default things that were added in, like maybe hidden love notes, bottles of extra-strength whiskey, monkey corpses, or low-end weapons like a knife, it'd be interesting.

nullbear commented:
I've only ever used this to hide soap.

Keep in mind that MOST of the time, to hide a powerful item, it means that you are SACRIFICING your TC or access to that item for the current round, with no guarantee whatsoever that you will be able to retrieve it.

I have a question though: What happens when the tile it's hidden under gets sucked up by a singularity or bomb? Is the satchel lost?

If someone hides a satchel -> It will appear probably several times over the course of the next 40 rounds, but almost definitely not in the next round when it would be 'desired' by someone who was attempting to powergame it.

I expect that the main cause of satchels being 'found' is by bombs/singuloth destroying them, and their contents being lost before ANYONE can find them. This is more likely than someone stumbling across it, or even the 'hider' finding it in a future round. If the 'hider' cant find it in the next 2-3 rounds they play (and it's likely they won't) then they'll probably stop looking for it, never obtaining it when it does appear, and forgetting about it when the pool gets heavy.

Iamgoofball commented:
also the problem with there never being anything seeded is the station gets blown the fuck up/singulo'd every other round

Incoming5643 commented:
Satchels are only lost if they happen to exist when the tile is destroyed, which means that at most one old satchel can be lost even if a singulo eats everything. The problem this pull had was that the equilibrium was set to 20 satchels, when 20 was the sheer minimum for anything to happen. Hence the rebalance pull.

@nullbear You don't have to sacrifice the item outright, just hide it after the round is nearly over on your way to the shuttle
